fdbserver/TLogGroup.h

```diff
@@ -0,0 +1,98 @@
+/*
+ * TLogGroup.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FDBSERVER_TLOGGROUP_H
+#define FDBSERVER_TLOGGRORUP_H
+
+#include "fdbrpc/Locality.h"
+#include "fdbserver/WorkerInterface.actor.h"
+#include "flow/FastRef.h"
+#include "flow/IRandom.h"
+#include "flow/network.h"
+#include <unordered_map>
+#include <vector>
+
+#pragma once
+
+// Represents a single TLogGroup which consists of TLog workers.
+struct TLogGroup {
+	TLogGroup() : groupId(deterministicRandom()->randomUniqueID()) {}
+
+	const UID& id() const { return groupId; }
+
+	void addServer(const NetworkAddress& serverAddr) { servers.push_back(serverAddr); }
+
+private:
+	UID groupId;
+	std::vector<NetworkAddress> servers;
+};
+
+struct TLogWorkerData {
+	UID id;
+	NetworkAddress address;
+	LocalityData locality;
+
+	static TLogWorkerData fromInterface(const WorkerInterface& interf) {
+		return TLogWorkerData{ interf.id(), interf.address(), interf.locality };
+	}
+
+	bool operator==(const TLogWorkerData& other) const {
+		// TODO: Is NetworkAddress enough?
+		return other.id == id;
+	}
+};
+
+// User-defined hash function for TLogGroup::WorkerData.
+namespace std {
+template <>
+struct hash<TLogWorkerData> {
+	std::size_t operator()(const TLogWorkerData& w) const noexcept { return w.id.hash(); }
+};
+
+template <>
+struct hash<TLogGroup> {
+	std::size_t operator()(const TLogGroup& w) const noexcept { return w.id().hash(); }
+};
+} // namespace std
+
+// `TLogGroupCollection` manages, recruits and tracks all the TLogGroups in the system.
+class TLogGroupCollection : public ReferenceCounted<TLogGroupCollection> {
+public:
+	explicit TLogGroupCollection(const Reference<IReplicationPolicy>& policy);
+
+	void addRecruits(const std::vector<WorkerInterface>& logWorkers);
+
+	void recruitEverything();
+
+private:
+	void rebuildLocalityMap(const std::vector<WorkerInterface>& logWorkers);
+
+	const Reference<IReplicationPolicy> policy;
+
+	LocalityMap<TLogWorkerData> localityMap;
+
+	std::unordered_set<TLogGroup> tLogGroupSet;
+
+	std::unordered_map<TLogWorkerData, UID> tLogMap;
```

Comment:
why is this needed? Do you mean worker to group UID map?

Anyway, can you add comments for all fields.